### PR TITLE
Enables LTO for Travis Release Builds

### PR DESCRIPTION
**Pull request by daniel-j-h**
LTO builds for Travis release jobs. Let's wait for Travis and see how it goes.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -58,7 +58,7 @@ matrix:
         apt:
           sources: ['ubuntu-toolchain-r-test']
           packages: ['libstdc++-5-dev']
-      env: CLANG_VERSION='3.9.1' BUILD_TYPE='Release' ENABLE_MASON=ON ENABLE_SANITIZER=ON
+      env: CLANG_VERSION='3.9.1' BUILD_TYPE='Release' ENABLE_MASON=ON ENABLE_SANITIZER=ON ENABLE_LTO=ON
```
Comment (TheMarex) [inline .travis.yml]: Would omit this one, I don't think need we need this for the sanitzer build.

Comment (daniel-j-h) [inline .travis.yml]: Hm having the same build setup sanitizes that we're using for releases would be nice though - without lto things like undefined behavior due to e.g. ODR violations can behave differently. I don't have a strong opinion here, though. Removed it for now.

**Comment by daniel-j-h** [comment]
Urgh I've just seen the Clang error with LTO builds on Travis

> clang-3.9: error: unable to execute command: Segmentation fault (core dumped)
> clang-3.9: error: linker command failed due to signal (use -v to see invocation)

**Comment by TheMarex** [comment]
@daniel-j-h weird, we are using clang 3.9.1 in node-osrm as well and we use it to compile `osrm-backend`. Any specifics about our setup?

**Comment by daniel-j-h** [comment]
Hm no, we are using _the exact_ same Clang version in node-osrm:
https://github.com/Project-OSRM/node-osrm/blob/master/scripts/build.sh#L17

The failing build is Mason+Clang3.9.1+LTO specifically. I can try to reproduce locally.

**Comment by springmeyer** [comment]
Thanks for the ping. I recall that clang 3.9.0 and binutils 2.27 had a bad interaction that was fixed in clang 3.9.1: https://github.com/Project-OSRM/osrm-backend/pull/3389#issuecomment-264247163. The fact that this is happening now with 3.9.1 is curious. I'll attempt to replicate later today and see if I see any clues.

**Comment by daniel-j-h** [comment]
I just spawned up a local Trusty Docker container, replicating the build environment with binutils, clang and really all deps. from Mason - except for zlib and stdlibc++ (from ubuntu toolchain ppa, as we do on Travis).

But I can not replicate the segfault.

What I tried is clearing the Travis caches primarily for ccache and restarting the build.
But we still hit the Clang 3.9.1 segfault.

**Comment by daniel-j-h** [comment]
It works with Clang 4.0. Argh. At least we are unblocked by it now.

@springmeyer as soon as we merge this PR we depend on Clang 4.0 for all our production builds. Can you make sure to bundle a recent Clang 4.0 RC and once the release is out (should be a matter of days) bundle this.

**Comment by daniel-j-h** [comment]
Blocked by having LLVM 4 in Mason https://github.com/mapbox/mason/issues/385

**Comment by springmeyer** [comment]
> Blocked by having LLVM 4 in Mason mapbox/mason#385

llvm 4.0 packages are now available via mason. This includes separate packages:

  - llvm (all tools in one download)
  - clang++ - just the compiler + libraries needed for `-fsanitize=*` and `-flto`
  - llvm-cov
  - clang-tidy
  - clang-format
  - lldb
  - include-what-you-use

**Pull request merged**